### ISSUE TITLE
Proposal: Remove Initial Configuration and Representation Declaration

### ISSUE THREAD

**Issue by bradrn** [catseye/ALPACA#9]
I am currently creating an application to simulate arbitrary cellular automata. To specify these automata, at first I used an improvised, ad-hoc, ugly language to specify the states, transitions and conditions. It was somewhat hard to write and almost impossible to read and modify. When I discovered ALPACA, it seemed like a good fit for my application, so I created a small interpreter and tested how ALPACA worked with my application. ALPACA turned out to be amazingly well designed, so much so that my application now uses ALPACA as its main specification language. It is easy to read, easy to write, and has many features which directly relate to its stated purpose as 'a language for defining cellular automata'.

In fact, all features of ALPACA relate to this stated purpose - except two. The initial configuration - and, by extension, the representation declaration - don't seem to have anything to do with it. This is less salient when considering the reference implementation, which as a console application is of course more static (that is, you can't edit a pattern after it's been loaded into the application) and text-based, but glaringly obvious if you use ALPACA with a GUI containing an editable grid, as I have been for the past month. The initial configuration seems to be particularly useless - there doesn't seem to be much use for it outside test cases and small examples. Other than these two applications, I cannot think of any situation where I would want to specify a playfield which is used by default every time a certain cellular automaton is invoked. As a consequence of this, the representation declaration doesn't seem particularly useful either, as the documentation defines it very explicitly as existing only to assist the initial declaration, and not for e.g. display purposes:

> ... if an initial configuration is included in the ALPACA description, the characters in it will be mapped to states by using the representation declaration of each state. Therefore, if a state has no representation declaration, no cell in that state can be present in the initial configuration.
> [...]
> Informational: an implementation may use the representation declaration of a state to display (or otherwise communicate) that state to the user. This is certainly a reasonable choice, and one expects many implementations will at least offer that as an option. However, representation of states is ultimately implementation-defined. Even if representation declarations are included for every state in the ALPACA description, there is nothing to prevent an implementation from overriding them with some other representation.

My proposal, therefore, is simple: **to remove the initial configuration and representation declaration**. The initial configuration seems pretty useless to me, and its removal would only really affect the Falderal test suite and some of the examples. However, the representation declaration is currently used for display purposes in the reference implementation, and thus its removal would be far more severe. I therefore suggest the implementation of the following suggestion you have made at the bottom of the documentation:

> Informational: It is expected that a future version of ALPACA Stylesheets will define a property called `glyph` ... with the understanding that it will be used primarily for display, while the Representation Declaration will be used primarily for interpreting a cellular automaton's configuration from a textual source, e.g. the initial playfield.

(Which incidentally makes it even more clear that the representation declaration exists only to interpret the initial playfield, and should thus be removed with it.)

**Comment by OrangeNote** [comment]
**In short: I think that ALPACA (the language) is strongly linked to its implementation.**

I'm also developing an application to design cellular automata, with ALPACA as the ideal foundation. But of course I had to change the representation from characters to colors, being on a graphical canvas, and I moved the initial representation away from the rest of the specification. So I understand your point of view, because I encountered the same inconsistencies.

But, even if it seems the most logical thing to do, don't forget that ALPACA is meant to be "runnable"
> [...] allowing a pre-defined CA configuration to be included with the CA description. (This last enhancement makes ALPACA CA-complete [...]

For what I can see, the author intent is to make ALPACA a language to all effects, more specifically a meta-language, capable of reading and modifying itself. From this point of view, the initial representation is crucial and can't be removed.

So why change the specs and/or the implementation of ALPACA, when you can make a modern adaptation of it, and removing the parts of the language that are closely related to its output interface (the console)? I'm curious to know what brings you to suggest a change to the original description, instead of proposing a new alternative.

_PS. I'm not the author of ALPACA nor a language expert but I do care about the topic and I think it's worth taking part in the discussion._

**Comment by bradrn** [comment]
Your point is valid @OrangeNote - I did indeed consider this statement about CA-completeness while I was thinking about this issue. However, the way I see it, the 'runnability' of ALPACA is a secondary goal. As its name suggests: A Language for the Pithy **Articulation** of Cellular Automata. I interpret this name as saying that the purpose of ALPACA is to allow the easy definition of cellular automata. The documentation concurs - in its first line, it states:

> This document describes version 1.1 of ALPACA, **a language for defining cellular automata**. _[my emphasis]_

The initial configuration is entirely tangential to this goal; its purpose is to define a playfield for the automaton to run on - a goal which is entirely different to ALPACA's stated purpose of 'defining cellular automata'.

I notice also that no-one has said anything yet about the second half of my proposal - the removal of the representation declaration. The representation declaration is currently used for two purposes only:

1. To specify the characters used in the initial configuration;
2. To display the playfield as it evolves.

The first point will be redundant if the initial configuration is removed. The second point is interesting - the documentation explicitly states that this is not the intended use of the representation declaration:

> Informational: an implementation may use the representation declaration of a state to display (or otherwise communicate) that state to the user. This is certainly a reasonable choice, and one expects many implementations will at least offer that as an option. However, representation of states is ultimately implementation-defined. Even if representation declarations are included for every state in the ALPACA description, there is nothing to prevent an implementation from overriding them with some other representation. See also the section on ALPACA Stylesheets, below.

So this use of the representation declaration is very much tied to the nature of the reference implementation, and is not really part of the specification. In general, I am of the opinion that the ALPACA language itself should not include any features which are specific to one implementation. If the initial configuration is removed, the _only_ reason to include the representation declaration would be to support the reference declaration - so it should be removed with the initial configuration.

**Comment by cpressey** [comment]
From https://github.com/catseye/ALPACA/issues/9#issue-323196358 :

> the representation declaration is currently used for display purposes in the reference implementation, and thus its removal would be far more severe

But this is a limitation of the reference implementation (and to some degree how the examples are presented in the spec), not the language. The spec is (I think) quite clear that representation declarations are an optional convenience and, even when they're supplied, they're not required to be used when rendering a playfield.  And since they're optional anyway, I don't see a pressing reason to eliminate them entirely.

ALPACA is not intended to be tied to any particular implementation - that's why it's got a spec. If the spec could stand to be clearer about what does and what does not define ALPACA, that's fair.

(Part of the problem is, when you have 2 normative definitions of a language (like, a spec and a reference implementation), how do you interpret it when they do things differently?  There is an argument here that you should have *3* normative definitions, so that there is a tie-breaker, and majority rules. But I digress.)

My latest ideas, based on this issue, run along these lines:
- remove initial configuration from ALPACA spec; result is Core ALPACA.
- define an ALPACA Container Format that allows a Core ALPACA definition and an initial configuration to be bundled together into a single file.
- bonus: the ALPACA Container Format can *also* allow bundling in an ALPACA Stylesheet!
- the reference implementation currently implements several features that go above and beyond the spec, and thus aren't really reference for anything.  It should be split into a program that implements only what's in the spec, and attempts to do so normatively, and a second implementation which includes the bells and whistles like compiling to Javascript and so forth.

**Comment by bradrn** [comment]
@cpressey - that's exactly what I was trying to say in https://github.com/catseye/ALPACA/issues/9#issuecomment-389387347! However, you don't need to worry about the spec being unclear - it is already very precise about what is part of the spec vs the reference interpretation, to the point that much of my argument was simply quoting the relevant parts of the spec.

I quite like your ideas, especially about the 'ALPACA Container Format'. However, I don't think your last point (about splitting the reference implementation) is necessary - it's already abundantly clear what is 'reference' and what is not.

**Comment by bradrn** [comment]
I've been thinking about this statement from @cpressey's earlier comment https://github.com/catseye/ALPACA/issues/9#issuecomment-389243177:

> it should be entirely possible [...] for [the initial configuration] to be detached from the "core" language spec, and made into an optional "module", much like ALPACA Stylesheets already is.

I've just realized that in the reference implementation we already have a 'modularised' initial configuration - we just don't realise it! Have a look at part of the output of `alpaca -h`:

```
  -i FILENAME, --initial-configuration FILENAME
                        initial configuration to load into playfield (when
                        evolving a playfield only)
```

Since we already have this in the reference implementation, it should be possible to just drop the initial configuration and just use this instead. Unfortunately, it isn't as simple as this, since I also wanted to drop the representation declaration, which is the mechanism which indicates which state each character corresponds to. In more detail, here is my proposed design:

> The new ALPACA language would have no initial configuration, and hence no representation declarations (see https://github.com/catseye/ALPACA/issues/9#issuecomment-389387347 for my reasoning behind this) - e.g. the Life example from the front page would become:
> ```
> state Dead
>   to Alive when 3 Alive and 5 Dead;
> state Alive
>   to Dead when 4 Alive or 7 Dead.
> ```
> For backwards compatibility, the definition would still end with a `.` character. It might also be a good idea to deprecate instead of remove the initial configuration, again for backwards compatibility.
> 
> The new 'initial configuration module' would be based on the file passed to `-i`. However, as I said above, 'it isn't as simple as this, since I also wanted to drop the representation declaration, which is the mechanism which indicates which state each character corresponds to'. In order to resolve this, I propose adding a section before the configuration itself, as follows:
> ```
> " "=Dead
> "*"=Alive
> begin
> ***
> * *
> * *
> ```
> So a section where character aliases can be defined, then `begin`, then the initial configuration proper.
> 
> Since the representation declarations have been removed/deprecated from Core ALPACA, for display purposes an extension to the stylesheet is required in the form of a `glyph` property:
> ```
> .Dead {
>     fill: #ffffff;
>     glyph: " ";
> }
> .Alive {
>     fill: #000000;
>     glyph: "*";
> }
> ```

Now that we have three separate files for one ALPACA description, @cpressey's idea of an ALPACA Container Format will be useful. I'm not entirely sure about how to design this format - does anyone else have any ideas?

**Comment by OrangeNote** [comment]
I think your proposed initial configuration module has a weakness: even if CA definition and initial configuration are separated, you're still using StateID as a key to map the representation to the state definition. And that's a problem because it means you can't reuse the initial configuration module with different CAs, unless they have the same StateIDs. Being modular is not an advantage in this sense, but I'm not sure if if a broader modularity (that is the possibility to reuse the representation with many different CAs without the need to remap states) is useful in real cases.
Actually, there's a similar problem with stylesheets, you can't use them as modular, reusable themes. Again, I don't know it's a feature which is worth discussing, even because it would be unclear how the mapping would work without an explicit identifier.

**Comment by cpressey** [comment]
@bradrn 

> However, I don't think your last point (about splitting the reference implementation) is necessary - it's already abundantly clear what is 'reference' and what is not.

I disagree, actually; I think a lot of what I'm reading in this thread sounds like it is clouded by the fact that the reference implementation does some thing in some way, and it's difficult to tell if doing that thing in that way is actually canonical (or not), so we assume it is (or isn't) normative.

This is exacerbated by the fact that the spec contains example programs that are intended to be run by the reference implementation. The spec relies on (or appears to rely on) the reference implementation, in a way that specs often don't do, and it would pay to be careful about that.

For example, yes, the reference implementation supports `--initial-configuration`.  Implementations are allowed to do that.  They're also not required to do that.  I don't see the relevance of that command-line option to the design of the language, so I find the reasoning in https://github.com/catseye/ALPACA/issues/9#issuecomment-389790432 difficult to follow.

I also think we may simply be overthinking this.  Even today, if you implement ALPACA in some tool of your own, but don't implement the initial configuration part, you could say,

> This program implements ALPACA 1.1 except for giving an initial configuration, which is not supported.

and I can't really see why that wouldn't be sufficient for everyday use?  Most CA definitions out in the world today that are written in ALPACA don't include initial configurations, and if one does, you can simply remove or ignore the initial configuration part if it's not relevant to you.

Sure, it would be nice if the spec was a bit clearer on this and it could be officially sanctioned, but it's not like the ramifications of not adhering precisely to the spec on this point are drastic (unless they are in your case, in which case explaining why would definitely help change my mind on the matter.)

@OrangeNote 

> Actually, there's a similar problem with stylesheets, you can't use them as modular, reusable themes.

Sadly, that seems largely true of CSS as well. I mean, CSS is modular to a degree, but front-end web developers are often coming up with new schemes for compensating for the problems you run into after you pass a certain point. (BEM comes to mind.)

(Continuing the CSS analogy: representation declarations are like inline styles, which are considered bad web development practice, but still supported by HTML, and sometimes useful for when quick-and-dirty work is OK.)

ALPACA Stylesheets weren't primarily intended to be modular; they were primarily intended to separate presentation from semantics.  However, if it possible for them to be a "a little modular", that's great.  (One example might be closely related CA's, like Jaccia and Jacciata, which have almost the same set of states; you can use just 1 stylesheet to style both of them, because they have some names and meanings in common.)

I haven't really considered the implications of treating initial configurations as "modular" either; but I suppose that might work in at least the same case, of closely related automata, i.e. there are some configurations which work analogously in both Jaccia and Jacciata.

At any rate, I don't think anyone should expect completely unrelated CAs to be able to share stylesheets, or configurations, in any meaningful way.  But I am fine with that.

...

I don't have a whole lot of time to dedicate to this but I'll think about what a container format might look like.

**Comment by bradrn** [comment]
@cpressey:

> [...] I think a lot of what I'm reading in this thread sounds like it is clouded by the fact that the reference implementation does some thing in some way, and it's difficult to tell if doing that thing in that way is actually canonical (or not), so we assume it is (or isn't) normative.

Here is my argument, in an attempt to divorce it from the reference implementation:

1. The ALPACA spec states that an implementation must implement an 'initial configuration' in order to be a compliant implementation of the language.
2. The initial configuration doesn't seem like a particularly useful feature; as I wrote earlier, 'there doesn't seem to be much use for it outside test cases and small examples'. It adds complexity to the parser and interpreter without adding value to the overall language.
3. Furthermore, in an interactive GUI system, the user can edit, load and save configurations easily, thus further reducing the need for an initial configuration.
4. Thus I propose the removal of the initial configuration from the specification.

> This is exacerbated by the fact that the spec contains example programs that are intended to be run by the reference implementation. The spec relies on (or appears to rely on) the reference implementation, in a way that specs often don't do, and it would pay to be careful about that.

I don't agree with this - the given examples don't seem to have anything to do with the reference implementation.

> I don't see the relevance of that command-line option to the design of the language, so I find the reasoning in https://github.com/catseye/ALPACA/issues/9#issuecomment-389790432 difficult to follow.

You're quite right - I should have realised this before I posted it. Feel free to disregard that comment.

---

Now, some thoughts of my own:

Currently I am implementing an encoder and decoder for the RLE file format, which has become the _de facto_ pattern exchange format for cellular automata. While doing this, I came to the realisation that almost all common pattern exchange formats (listed e.g. [here](http://psoup.math.wisc.edu/mcell/ca_files_formats.html)) depend on some sort of correspondence between cell states and characters. In https://github.com/catseye/ALPACA/issues/9#issuecomment-389790432, I achieved this by adding a section to the file format to achieve this, but obviously you can't just add a section to a well-established format such as RLE or MCell. The solution is simple - use the representation declaration! I was quite adamant earlier on that the representation declaration should be removed, but now that I've discovered a good reason to keep it I've changed my mind on this matter.

**Comment by bradrn** [comment]
Yes, I realise it's a bit strange to say I'll do something 'as soon as I can' and then get it done it three days later. There was unfortunately a bit of other work I had to do which prevented me from getting to this. It also probably didn't help that I kept on changing my mind about what I wanted to say next...

So anyway, here's the list of pros and cons to the initial configuration, ordered from most important (in my mind) to least important:

> Pros:
> 1. If you do need a fixed initial configuration, you can use it
> 2. Useful with console applications (e.g. the reference implementation)
> 3. @OrangeNote: "The author's intent is to make ALPACA a language to all effects"
> 
> Cons:
> 1. Separation between description and presentation
> 2. Orthogonal to the stated aim: 'A language for defining cellular automata'
> 3. Doesn't have many uses
> 4. @cpressey: Not a 'core' part of the language
> 5. Doesn't work well with GUIs (e.g. the implementation I'm developing)

Now, let's have a look at these in more detail. To me, pro 1 and cons 1 and 2 (they're sort of the same anyway) seem quite important; I would say that the other points are much less convincing. First of all, pro 1 I would say is the most important point of all: whatever the eventual outcome of this issue is, the initial configuration _should_ be kept in some way, to ensure that if there does happen to be a need for the initial configuration it can still be used. However, I would say that cons 1 and 2 are also important. Ideally, the initial configuration could be moved into a separate file (similar to ALPACA stylesheets) but remain part of the specification.

Hopefully this clears up some of the confusion there has been around this issue.
